docs/developer-docs/production/storage.md

```diff
@@ -0,0 +1,112 @@
+
+You can learn more about storage costs [here](/docs/current/developer-docs/gas-cost).
+
+## Storage handling in different languages
+
+### Motoko storage handling
+
+In Motoko canisters, stable memory can be utilized through the Motoko **stable storage** and **stable variable** features. Stable storage is a Motoko-specific term used to refer Motoko's implementation of stable memory used to persist data across canister upgrades. Stable variables are Motoko defined variables that use the `stable` modifying keyword to indicate that the variable's value should persist across canister upgrades, such as:
+
+```motoko
+actor Counter {
+
+  stable var value = 0;
+
+  public func inc() : async Nat {
+    value += 1;
+    return value;
+  };
+}
+```
+
+To utilize stable memory when upgrading a Motoko canister, the following workflow can be used:
+
+- The canister must be stopped.
+- A `pre_upgrade` hook can be executed if one is defined. `pre_upgrade` hooks can be used for functions such as creating a backup of the canister's data. 
+- The canister's heap memory is discarded and a new version of the canister's Wasm module is initialized. Data stored in stable memory is made available to the new Wasm module.
+- The canister's new code is installed using the `--mode upgrade` flag.
+- The canister is started and now runs the newly upgraded code.
+- Stable variables are re-loaded as part of a `post_upgrade` hook. Stable memory bytes that stored those variables are overwritten with zeroes to minimize stable memory costs for the canister.
+
+#### Garbage collection
+
+A garbage collection process is an automatic utility that is used to manage the amount of memory used. For heap memory, garbage collection is used to remove unreferenced or dead objects in order to free up otherwise allocated heap memory. 
+
+In Motoko, the default garbage collection process uses a copying approach, which is depends on the size of the amount of heap memory currently used. In contrast, an additional garbage collector that uses a marking approach can be selected, which is based on the amount of free heap memory. These garbage collection methods are triggered when there are enough changes made to heap memory since the last round of garbage collection. Garbage collection can be forced to run after every message using the `--force-gc` flag in the project's `dfx.json` file:
+
+```json
+  "defaults": {
+    "build": {
+      "packtool": "",
+      "args": "--force-gc"
+    }
+  },
+```
+
+Both of these garbage collectors, however, are unable to garbage collection the entire heap memory pool due to the instruction limit per message on ICP. Since these garbage collectors cannot collect the entire heap memory pool, canisters do not benefit from fully utilizing the entire 4GiB memory pool, as there must be some heap memory space left for the garbage collector to operate.
```

Comment:
```suggestion
Both of these garbage collectors, however, are unable to collect the entire heap memory pool due to the instruction limit per message on ICP. Since these garbage collectors cannot collect the entire heap memory pool, canisters do not benefit from fully utilizing the entire 4GiB memory pool, as there must be some heap memory space left for the garbage collector to operate.
```